scipy/stats/tests/test_stats.py

```diff
@@ -7209,10 +7223,20 @@ def test_combine_weights(self):
                 [1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1]),
             stats.wasserstein_distance([5, 0, 1], [0, 4, 3],
                                        [1, 2, 4], [1, 2, 4]))
```

Comment:
If the test above were under review today, I would have suggested:
```python3
rng = np.random.default_rng(4437304399483663605)
n_values, max_repeats = 10, 10
repeats = rng.integers(0, max_repeats, size=n_values)
u_values = rng.random(size=n_values)
u_weights = repeats * rng.random()  # overall scale of weights doesn't matter
v_values = rng.permuted(np.repeat(u_values, repeats))
v_weights = np.ones_like(v_values) * rng.random()
res = stats.wasserstein_distance(u_values, v_values, u_weights, v_weights)
assert_allclose(res, 0, atol=1e-15)
assert np.any(repeats == 0)  # true even for an edge cases like this
```

This is a property-based test that emphasizes the relationships that must be true without the distraction of the specific numerical values involved. It's easier for me to review because I only have to review the code rather than dozens of individual numbers. It ensures that the values weren't hand-selected to make the test pass, and the tests written like this tend to be stronger because it's easy to scale the size of the test and the number of instances tested. Note that I also tested that the total weight doesn't matter, fractional weights work as we'd expect, and I checked an edge case (one of the weights is zero, which is unusual).

`test_shift_nD` and `test_rotation_reflection_nD` already seem to be written like this. Great! Would you mind converting the other tests to property-based tests, since they are all checking relationships, not something to do with the particular numbers involved? If there are cases that don't lend themselves well to this style of test, or if the code would be much more complex than just hard-coding values, let me know, but I'm guessing that most will be amenable.

Another property-based test I'd like to see, if it is not already here (just point it out if it is): applying `wasserstein_distance` to 2D column arrays (and thus calculating the result with `milp`) gives the same distance as applying it to the flattened arrays (so that the distance is computed with `_cdf_distance`.

Reply:
Thanks. I can get most of your points but can we clarify a bit about the code you provided? 
The last line
``` python
assert np.any(repeats == 0)  # true even for an edge cases like this
```
raise an Assertion error if the `repeats` variable has no zero values, right? Does this has any relationship with the `wasserstein` function behavior?

>Another property-based test I'd like to see, if it is not already here (just point it out if it is): applying wasserstein_distance to 2D column arrays (and thus calculating the result with milp) gives the same distance as applying it to the flattened arrays (so that the distance is computed with _cdf_distance.

I think `test_multi_dim` is what you would like to have.

Reply:
> raise an Assertion error if the repeats variable has no zero values, right? I don't see how the wasserstein function behavior can affect this.

Correct in both cases. Another way to do that is to manually add a zero to `repeats`. (Would that be better practice? I'm not sure.) Instead, since a zero is already there with the random seed I happened to generate, I chose to just highlight that they are already there.

You're right that `wasserstein_distance` can't affect the pass or failure of that line. The line is simplify verifying a feature of the test itself. If it fails due to a change of the test, then we are alerted that the test no longer covers this edge case.

Reply:
Okay I see.
> since a zero is already there with the random seed I happened to generate, I chose to just highlight that they are already there.

but can you still call this a 'property based test' when it actually depends on your specific random seed :rofl: ?

Reply:
I'm not sure if that was rhetorical, so I'll answer. The test of `wasserstein_distance` itself is property-based. Precisely what the test checks depends on the seed and structure of the test, which in some sense could be said of any property-based test. If you wanted to add a stochastic but non-zero number of zeros to `repeats` at random positions, you could do that to make this feature of the test less dependent on the seed. This doesn't seem necessary to me because for most seeds there will happen to be at least one zero already. Also, the presence of a zero is not an essential element of the test - just something extra I noticed that the test already covered by the existing seed, which I generated at random. I hope you agree that either way there is a big improvement over what is there now.

Reply:
I agree that either way can largely improve the current test. Yet I'm still questioning it because I'm not convinced that the zero values in a vector should be considered as a corner case. First the zero value is not special to `test_combine_weights` as they are very normal inputs like other positive and negative inputs (negative inputs are legal as long as their summed weight on a single value is positive and finite), and we already have another test `test_zero_weight` which is designed specifically for the zero weight case, so I would suggest we remove the last line of none zero vector assertion for simplicity and comprehensibility.

Reply:
Also, I believe your proposed changes should go to `test_same_distribution`.

Reply:
Yeah agree.